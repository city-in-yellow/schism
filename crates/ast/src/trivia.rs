#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub struct Range {
    pub start: usize,
    pub end: usize,
}

#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct Trivia<T> {
    pub inner: T,
    pub span: Range,
}

pub fn new<T>(inner: T, span: Range) -> Trivia<T> {
    Trivia { inner, span }
}

pub trait WithTrivia {
    fn pretty_string(&self, indent: usize) -> impl AsRef<str>;
}
