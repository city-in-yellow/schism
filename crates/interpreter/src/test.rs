use ast::trivia::WithTrivia;
use indoc::indoc;
use insta::assert_snapshot;

use crate::run_on_src;

fn e(s: &str) -> String {
    match run_on_src(s) {
        Ok(o) => {
            format!("{}\n", o.pretty_string(0))
        }
        Err(e) => e.to_report(s),
    }
}

#[test]
fn test() {
    assert_snapshot!(e("1"));
    assert_snapshot!(e("()"));
    assert_snapshot!(e(r#""hi there""#));
    assert_snapshot!(e(r#"foo is 1. foo"#));
    assert_snapshot!(e(r#"[1, 2, 3]"#));
    assert_snapshot!(e(r#"foo bar is bar. foo 1"#));
    assert_snapshot!(e(r#"1 + 2"#));
    assert_snapshot!(e(indoc! {r#"
        foo bar baz is
            barbaz is bar + baz.
            bazbar is bar * baz.
            bazbar - barbaz.
        
        foo 3 5
    "#}));
    assert_snapshot!(e(r#"[1, 2, 3] + 12"#));
    assert_snapshot!(e(r#"1 + "2""#));
    assert_snapshot!(e(indoc! {r#"
        foo bar baz bil is
            barbaz is bar + baz.
            barbaz + bil.
        
        foo 3 5 "a"
    "#}));
    assert_snapshot!(e(r#"10 - 4"#));
    assert_snapshot!(e(r#"0 - 12"#));
    assert_snapshot!(e(r#"1 - [1, 2, 3]"#));
    assert_snapshot!(e(r#""hello" - 1"#));
    assert_snapshot!(e(
        r#""this is a test of our regularly scheduled broadcast" * (0-1)"#
    ));
    assert_snapshot!(e(r#"5 / 2"#));
    assert_snapshot!(e(r#"(1 / 10) + (2 / 10)"#));
    assert_snapshot!(e(r#"1 > (1 / 10)"#));
    assert_snapshot!(e(r#"1=1 or 1<=0"#));
    assert_snapshot!(e(r#"3...6"#));
    assert_snapshot!(e(r#"3..=6"#));
}
