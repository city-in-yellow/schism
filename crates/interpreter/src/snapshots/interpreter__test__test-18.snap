---
source: crates/interpreter/src/test.rs
expression: "e(r#\"(1 / 10) + (2 / 10)\"#)"
---
0.3 0..18

